tests/unit/test_storage.py

```diff
@@ -141,60 +139,3 @@ def test_reduced_chunk_fully_masked_data_vmax():
                          method=np.mean)
     assert rc[0].size == 0
     assert rc[1] is None
-
-
-def test_like_s3_reduce_chunk():
-    """Test reduce chunk entirely."""
-    rfile = "tests/test_data/cesm2_native.nc"
-    offset = 2
-    size = 128
-
-    # no compression
-    rc = st.reduce_chunk(rfile, offset, size,
-                         compression=None, filters=None,
-                         missing=[None, None, None, None],
-                         dtype=np.dtype("int32"), shape=(32, ),
-                         order="C", chunk_selection=slice(0, 2, 1),
-                         method=np.min)
-
-    assert rc[0] == 134351386
-    assert rc[1] == 2
-
-
-def test_s3_reduce_chunk():
-    """Unit test for s3_reduce_chunk."""
-    rfile = "tests/test_data/cesm2_native.nc"
-    offset = 2
-    size = 128
-
-    # no compression, filters, missing
-    # identical test to one from test_storage
-    object = os.path.basename(rfile)
-
-    # create bucket and upload to Minio's S3 bucket
-    S3_ACTIVE_STORAGE_URL = "http://localhost:8080"
-    S3_URL = 'http://localhost:9000'
-    S3_ACCESS_KEY = 'minioadmin'
-    S3_SECRET_KEY = 'minioadmin'
-    S3_BUCKET = 'pyactivestorage'
-    # local test won't build a bucket
-    try:
-        test_bigger_data.upload_to_s3(S3_URL, S3_ACCESS_KEY, S3_SECRET_KEY,
-                                      S3_BUCKET, object, rfile)
-    except:
-        pass
-
-    # test on local machine fails; should pass when connected to Minio
-    try:
-        tmp, count = s3_reduce_chunk(S3_ACTIVE_STORAGE_URL, S3_ACCESS_KEY,
-                                     S3_SECRET_KEY, S3_URL, S3_BUCKET,
-                                     object, offset, size,
-                                     None, None, [],
-                                     np.dtype("int32"), (32, ),
-                                     "C", [slice(0, 2, 1), ],
-                                     "min")
-        assert tmp == 134351386
-        # count is None; no missing data yet in S3
-        assert count == None
-    except:
-        raise
```

Comment:
I don't want to assume anything about how much you've read about different types of software testing, so please ignore me if I'm teaching grandma to suck eggs, but here's my understanding of some differences between unit and integration tests:

unit tests:
* test a small unit of code in isolation (module, function, etc)
* may test internal code
* have no external dependencies
* are easy to run and run very quickly (quick enough to run as a commit hook)
* may try to simulate some error conditions that may be difficult to trigger

These tests simply test the `activestorage.s3.reduce_chunk` function, which is internal to this library. I have mocked out the `activestorage.s3.request` function that performs the communication with the S3 active storage server, so that it does not require one to be running. An error condition is simulated where the server returns 404 not found, so that we can test the appropriate exception is raised.

On the other hand, integration tests:
* test the integration of multiple systems and/of software in an 'end to end' test
* rely on public interfaces
* may have complex dependencies
* may be difficult to set up and take a long time to run

So in our case this would involve integrating `PyActiveStorage` with some storage backend and optionally an active storage server. For the S3 case, this would be an S3 object store, and an S3 active storage server. For local file storage, it would just use local disk. The tests should use the public `Active` interface of `PyActiveStorage`. Over time we may build up various test cases including different data sets and configurations to exercise the system, and this integration testing may take a long time.

There's plenty of resources to read about it, but as usual SO does it well: https://stackoverflow.com/questions/5357601/whats-the-difference-between-unit-tests-and-integration-tests

Reply:
thanks a lot @markgoddard - both for contributing the tests, and for the detailed description of tests - I have a fair bit of experience with testing so you may have wanted to save your time on the above comment, but hey, one can never learn enough :grin: What I am after at this very moment is to establish a connection between PyActiveStorage, specifically the open file mechanism in it, and a netCDF file stored on an S3-type of storage, nothing else as pressing as that, unit/integration/-testing/cleaning up etc, these would follow. Life's made much easier for me now with your unit tests added - but I'll have to do some cleaning up since we can't really merge a PR with all those multiple commits of mine, it's pretty bad for git history. Next time you open a PR please base it off `main` - I keep the main branch as clean and as up to date as one can do :grin: :beer:

Reply:
Brilliant, cheers muchly :beer: